### ISSUE TITLE
Differing definitions of structs have unintended consequences

### ISSUE THREAD

**Issue by borrrden** [couchbase/couchbase-lite-core#22]
From [this SO question](http://stackoverflow.com/q/40731893/1155387)

Defining a struct one way for external use and another way for internal use can have unintended consequences on (at least) x86 platforms.  The summary is that the compiler will use the definition of the type to determine the way it should be returned from a function, and in particular if it needs special treatment with regards to exceptions (e.g. defined constructor) then it will be returned via memory while otherwise it will be returned via register if possible.  In the case of `FLSlice` and `C4Slice` this is happening, causing disagreement between the caller and callee about what return method to use.  The end result is a very hard to understand invalid memory access due to the callee looking in the wrong place for the parameter that was just passed because it expected space for the return value at the beginning when the caller did not leave any.  

There needs to be a refactor to the "internal / external" functions to switch between the two distinct types to avoid this.

**Comment by borrrden** [comment]
Pushed up some potential work on branch `issue/22` (confirmed that it solves the problem).  Also, the `internal` name of the function in `Document.hh` was causing clashing with other things called `internal` for some reason so I renamed it.

Here is the status:

FLSlice (and C4Slice) are now defined the same internally and externally except internally they have a few convenience operators.  I wasn't able to figure out how to get the FLSlice and C4Slice definition into slice.hh without breaking the whole build so those functions are separate but they both simply do a C-style reinterpret cast-copy 

```
FLSlice = *((FLSlice *)&x);
slice = *((slice *)&x);
```

These could both be `#ifdef`'d to no-ops if required and hopefully the compiler will just strip them all together if they are just returning the variable as is.
